### DISCUSSION TITLE
Docs: Clarify that first-party `@next/mdx` also has dynamic `import()` - not just `next-mdx-remote`

### DISCUSSION THREAD

**Discussion by karlhorky** [vercel/next.js#70417] [Ideas]
### What is the documentation issue?

Since a few months / years, I have encountered articles such as [MDX in Next.js - Space Jelly](https://spacejelly.dev/posts/mdx-in-nextjs) by @colbyfayock which are based on the (incorrect) assumption that `@next/mdx` cannot pull in content from other sources such as files outside of the `app` directory:

> The trick here though is that to use the official Next.js MDX package, our content needs to be located inside of the app directory, just like any other page file.

> But in our example, say we have a content directory in the root of our project with a dump of our MDX, and it’d be tedious, and frankly a bit overwhelming to put that all inside of the app directory with the page filename pattern, we can instead use Next MDX Remote to render our content from where it is.

This is an incorrect assumption, as files outside of the `app` directory can be pulled in via static imports and dynamic `import()`:

1. static `import`
   - https://nextjs.org/docs/app/building-your-application/configuring/mdx#using-imports
   - https://github.com/vercel/next.js/blob/3127e868d8950f4b78858c1ff4f40357487bcdb3/examples/app-dir-mdx/app/page.tsx#L4
2. dynamic `import()`
   - https://github.com/vercel/next.js/discussions/12151#discussioncomment-268425
   - @wesbos in https://github.com/vercel/next.js/discussions/51016#discussioncomment-10360255

Apart from the aforementioned blog posts, there seems to be a bunch of community confusion on what is possible using `@next/mdx`:

- https://github.com/vercel/next.js/issues/47523
- https://github.com/vercel/next.js/discussions/58575
- https://github.com/vercel/next.js/discussions/50897
- https://github.com/vercel/next.js/discussions/47621
- more...

Even [the Next.js MDX docs section "Remote MDX"](https://nextjs.org/docs/app/building-your-application/configuring/mdx#remote-mdx) seem to (incorrectly) implicitly indicate that `@next/mdx` cannot pull in content from other sources such as files outside of the `app` directory:

> If your MDX files or content lives somewhere else, you can fetch it dynamically on the server. This is useful for content stored in a separate local folder, CMS, database, or anywhere else. A popular community package for this use is [`next-mdx-remote`](https://github.com/hashicorp/next-mdx-remote#react-server-components-rsc--nextjs-app-directory-support).

`@next/mdx` can achieve the same things as [`next-mdx-remote`](https://github.com/hashicorp/next-mdx-remote) and [`mdx-bundler`](https://github.com/kentcdodds/mdx-bundler) and [`rsc-mdx`](https://github.com/zhangyu1818/rsc-mdx), without the [significant, non-obvious downsides with the alternatives](https://github.com/vercel/next.js/discussions/50897#discussioncomment-10531319):

- [`next-mdx-remote/rsc` doesn't allow for importing MDX files within MDX files](https://github.com/hashicorp/next-mdx-remote/issues/398#issuecomment-1999743798)
- [`mdx-bundler` requires complicated manual configuration to share components to imported `.mdx` files](https://github.com/kentcdodds/mdx-bundler/pull/227) (not to mention pulling in a bunch of extra bundling dependencies and running the bundling outside of the existing webpack / Turbopack bundler config)

cc @manovotny @leerob @wooorm @remcohaszing 

### Is there any context that might help us understand?

I'm hoping the context is included above. I can add more if needed.

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/app/building-your-application/configuring/mdx

**Comment by samcx** [comment]
@karlhorky Thank you for sharing your thoughts! I moved this over to discussions because this doesn't necessarily seem like an issue with incorrect information with our documentation.

I'm actually not 100% familiar with all the capabilities with `@next/mdx`. Did you want to possibly push a PR with changes to our MDX docs?

**Reply by karlhorky** [reply]
I think the closest thing to a "docs bug" is this part here from my post above - the docs are misleading users, implying that `@next/mdx` is not capable of dynamically importing MDX:

> Even [the Next.js MDX docs section "Remote MDX"](https://nextjs.org/docs/app/building-your-application/configuring/mdx#remote-mdx) seem to (incorrectly) implicitly indicate that `@next/mdx` cannot [pull in content from other sources such as files outside of the `app` directory]:
> 
> > If your MDX files or content lives somewhere else, you can fetch it dynamically on the server. This is useful for content stored in a separate local folder, CMS, database, or anywhere else. A popular community package for this use is [`next-mdx-remote`](https://github.com/hashicorp/next-mdx-remote#react-server-components-rsc--nextjs-app-directory-support).

This is quite possible with RSC and async functions and first-party `@next/mdx`, as [shown by Wes](https://github.com/vercel/next.js/discussions/51016#discussioncomment-10360255):

```tsx
export async function BlogPostPage(props) {
  const { default: MDXContent } = await import(`../../posts/${props.params.slug}.mdx`);

  // ...

  return (
    <div>
      {/* ... Other things ... */}
      <MDXContent />
    </div>
  );
}
```

**Reply by samcx** [reply]
@karlhorky 

```
const { default: MDXContent } = await import(`../../posts/${props.params.slug}.mdx`);
```

Not sure this is a pattern we should have in the documentation, as folks may get confused and run into Webpack issues, as shown in that thread. Have you also successfully dynamically imported an mdx file from the filesystem?

**Reply by karlhorky** [reply]
> as folks may get confused and run into Webpack issues, as shown in that thread

Why do you think that? I think people are confused because there is no example in the docs.

2 reasons why I think it's better than the current option: 

1. It's about the same complexity as other patterns suggested in the docs
2. It's less complicated than the current suggestion to use [`next-mdx-remote`](https://github.com/hashicorp/next-mdx-remote?tab=readme-ov-file#react-server-components-rsc--nextjs-app-directory-support)

> Have you also successfully dynamically imported an mdx file from the filesystem?

Yes, we switched over to this in 3 parts of our large app, and it's better than the other options we tried: `next-mdx-remote`, `mdx-bundler`, and manually parsing + evaluating with `node:fs` and lower level MDX packages.

To be clear:

A) It's a feature of the first-party `@next/mdx` which is not documented
B) It's easier to use than the existing suggestion in the docs
C) It has features which are either not possible or complicated to achieve with the existing suggestions in the docs

**Reply by wesbos** [reply]
Yeah I agree - I wasted probably a day on this. Went down the MDX-remote rabbit hole only to find out that images outside the public dir dont work. If we want all the benefits of next - like images - we need to use the baked in next.js approach.

Good docs on how to use it with dynamic content are really needed. It's a bit of a footgun in that webpack is weird with dynamic imports - but that should also be documented as a gotcha. Or at least a better error message when webpack pukes because it's been scanning 600,000 files in node_modules

**Reply by samcx** [reply]
@karlhorky @wesbos Thanks for the clarification!

I'll take a look into updating our documentation here about dynamic content. :blob_bowing:

**Reply by karlhorky** [reply]
Amazing, thanks! Feel free to tag me on the PR - happy to give it a review.

**Reply by karlhorky** [reply]
@samcx there was a discussion on Twitter today about Next.js MDX options, with some missing knowledge that `@next/mdx` is a possibility for content outside of the `app/` directory

- https://x.com/peduarte/status/1845814417839034738

this kind of nerdsniped me and I made a demo repo showing that it's definitely possible to `await import()` MDX files outside of the `app/` directory:

- https://github.com/karlhorky/next-mdx-dynamic-import-demo


https://github.com/user-attachments/assets/6f4abc8b-57ff-49ef-b9bc-a014ea0ff499

![2024-10-15_18-10](https://github.com/user-attachments/assets/c4892a12-433c-4d33-89b7-dce7c234572a)

**Reply by lauri865** [reply]
This is actually great. Why the docs don't showcase this is very strange indeed. I lost way too much time trying to create reusable templates with next/mdx. One problem remains though – mdxrs doesn't support remark/rehype plugins (no turbo support), and there's been little movements on that front for a while. Good luck doing any templating without frontmatter. Hence we ended up using content collections instead, which works great even with turbo, and offers nice typesafety as a bonus. And we could construct our mdx files in a way which would requires minimal (to no) refactoring in case we do want to host the mdx content remotely at some point.

If instead we would have started exporting metadata from the mdx files as the docs suggest, we would need to refactor every single mdx file in the future when switching to a remote.

At this point I wish next/mdx didn't exist at all, because it just sent us down a rabbit hole of various blockers and incompatibilities.

**Reply by karlhorky** [reply]
@samcx what do you think about the new updates in the thread here?

Would love to get this documented! 🚀

As I mentioned before, I'm happy to review a draft of any docs PR that you have to document this 👍 Can even collaborate on creating the material too - but probably cannot start the PR from scratch right now.

**Reply by samcx** [reply]
Let me know what you think! @karlhorky → https://github.com/vercel/next.js/pull/73466

**Reply by karlhorky** [reply]
@samcx Amazing, thanks! 🙌 I'll take a look and add review comments :)
